project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/model/item/index/CompositeIndex.kt

```diff
@@ -0,0 +1,30 @@
+package org.plan.research.minimization.plugin.model.item.index
+
+import org.plan.research.minimization.plugin.psi.stub.KtStub
+
+import arrow.core.raise.option
+import com.intellij.psi.PsiElement
+import fleet.util.indexOfOrNull
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtNamedFunction
+
+import kotlin.collections.filter
+
+sealed class CompositeIndex : PsiChildrenPathIndex {
+    data class ChildrenNonDeclarationIndex(val childrenIndex: Int) : CompositeIndex() {
+        override fun getNext(element: PsiElement): PsiElement? =
+            element.children.filter { it !is KtClass && it !is KtNamedFunction }.getOrNull(childrenIndex)
+
+        companion object {
+            fun create(parent: PsiElement, child: PsiElement) = option {
+                ChildrenNonDeclarationIndex(
+                    ensureNotNull(
+                        parent.children.filter { it !is KtClass && it !is KtNamedFunction }.indexOfOrNull(child),
+                    ),
+                )
+            }
+        }
+    }
+
+    data class StubDeclarationIndex(val stub: KtStub) : CompositeIndex(), PsiChildrenPathIndex by stub
```

Comment:
Is it really needed? I mean, what if we will make only ChildrenNonDeclarationIndex as a PsiChildrenPathIndex and get rid of CompositeIndex?

Reply:
One more reason why this will be better: the current solution won't work correctly within tries

Reply:
Basically, the main idea why ever the stubs are needed here is to point to function/property where it has been called. 

The stubs here are required for the same reason as we use them in the declaration stage: we can't just use children (ever non-declaration ones) because the declaration will be deleted and indexes will be invalidated.

However in the bodies for now we don't delete any elements expect constructor parameters. However, constructor parameters are in the subtree of the constructors which are the most deep nodes at this representation.

However, I need to test how it will be working with constructions like

```kotlin
A(B(C(D(1))))
```

Reply:
Got it, that changes everything